subprojects/antlr/src/main/java/org/gradle/api/plugins/antlr/internal/AntlrWorkerManager.java

```diff
@@ -18,15 +18,16 @@
 
 import org.gradle.api.file.FileCollection;
 import org.gradle.process.internal.JavaExecHandleBuilder;
+import org.gradle.process.internal.MemoryResourceManager;
 import org.gradle.process.internal.worker.SingleRequestWorkerProcessBuilder;
 import org.gradle.process.internal.worker.WorkerProcessFactory;
 
 import java.io.File;
 
 public class AntlrWorkerManager {
 
-    public AntlrResult runWorker(File workingDir, WorkerProcessFactory workerFactory, FileCollection antlrClasspath, AntlrSpec spec) {
-
+    public AntlrResult runWorker(File workingDir, MemoryResourceManager memoryResourceManager, WorkerProcessFactory workerFactory, FileCollection antlrClasspath, AntlrSpec spec) {
+        memoryResourceManager.requestFreeMemory(0);
```

Comment:
`WorkerProcessFactory` should take care of this. When `WorkerProcessFactory` is asked to start a worker process, daemon or otherwise, it should obtain a lease for memory for that worker based on the requested heap settings of the worker process. Then, when the worker completes, however that happens, it should give up this lease.